File: Code/Python/bellman_ford.py
'''
    Bellman Ford provides shortest path from a given source to all the vertices in the graph.
    Steps:
        1. Initialize a distance array d as inifinite and intialize the distance to the source as 0, d[0]=0.
        (Distance array stores the distance of the nodes to the given source)
        2. Relax all the edges for V-1 times. For every edge u-v :
            if d[u]+w<d[v]:
                d[v]=d[u]+w
            where d[u]= distance of vertex u from origin
            d[v] = distance of vertex v from origin
            w = weight of the u-v edge
        3. Relax all the edges for one more time and if d[u]+ weight of u-v edge < d[v] then that means that
        there is a negative weight cycle and bellman ford algorithm can't work on graphs with negative weight cycle.
'''

def bellman_ford(n,g):
        # distance array
        d=[float('inf') for i in range(n+1)]
        # initialising the distance of source as 0
        # source here is 1
        d[1]=0
        
        for i in range(n-1):
            for u,v,w in g:
                if d[u]+w<d[v]:
                    d[v]=d[u]+w
        # checking for negative weight cycle
        for u,v,w in g:
            if  d[u]+w<d[v]:
                print("Negative weight cycle")
                break
        else:
            # print the distance of all the vertices from the source
            for i in d[2:]:
                print(i, end=" ")
        return
    
for _ in range(int(input())):
      # n is no of vertices and m is the no of edges
      n,m= list(map(int, input().split()))
      # initialising graph as a list
      g=[]
      for _ in range(m):
          # u-v is a directed edge with v as a weight
          u,v,w=list(map(int, input().split()))
          g.append([u,v,w])

      bellman_ford(n,g)

'''
    Test cases:
    Input 1:
    1
    5 5
    1 2 5
    1 3 2
    3 4 1
    1 4 6
    3 5 5
    Output:
    5 2 3 7

    Input 2:
    1
    3 3
    0 1 -1
    1 2 -2
    2 0 -3
    Output:
    Negative weight cycle

    # time complexity: O(VE), where V is no of vertices and E is no of edges
    # space complexity: O(V+E) for storing adjacency list
'''

File: Code/C++/bellman_ford.cpp
/* 
# Bellman Ford's Algorithm:
Bellman Ford algorithm helps us find the shortest path from a vertex(root) to all other vertices of a weighted 
graph. It can be implemented using DP. It uses principal of relaxation Which states that assume the same result 
(infinity here) for all the elements (nodes here) from the root node beforehand and then tweak these values 
according to a certain condition.
Similarly, in Bellman ford's algorithm loop runs for (|v|-1) times and each iteration tweaks the distance value
according to the below condition:

if ( assumed distance of  node a+distance between node a and another node b connected by an edge < assumed distance of node b)
{
    updated distance of b =  assumed distance of  node a + distance between node a and another node b  connected by an edge
}
It is similar to Dijkstra's algorithm but it can work with graphs with negative weights.

# Time Complexity:
It is O(e*(v-1)) where v and e represents no. of vertices and edges respectively.
- For completed graph:
no. of edges = v(v-1)/2
because in connected graph, each vertex is connected with every other node.
So. time complexity will be (v(v-1)/2)v which is equivalent to O(v^3)
- For Graph incomplete graphs:
no. of edges = v
So. time complexity will be v(v-1) which is equivalent to O(v^2)
- Worst Time Complexity: O(v^3).

# Space Complexity:
It is O(v+e) and O(v^2) when using adjacency list and adjancency matrix respectively.

# Advantage Of Bellman Ford Algorithm:
- It can calculate the shortest path in the graphs having negative weight edges whereas Dijkstra can't.
- A very important application of Bellman Ford is to check if there is a negative cycle in the graph.

# Drawbacks of Bellman Ford Algorithm:
If a graph contains a cycle whose all edges sum is negative then Bellman Form Algorithm will give the
wrong result because that cycle will reduce the total path distance by coming back to the same point.
That's why Algo will keep on changing the distance values even after (|v|-1)th relaxation cycle .

# why would one even have edges with negative weights in real life?
Negative weight edges might seem useless at first but they can explain a lot of phenomena like 
cashflow, heat released/absorbed in a chemical reaction etc.
For instance, if there are different ways to reach from one chemical A to another chemical B, each method will have sub-reactions involving both heat dissipation and absorption.

Implementation:
*/

#include<bits/stdc++.h>
using namespace std;
const int EDGES = 2e5;
const int NODES = 1e5;
const int MAX = 1e9;
vector<int> v[EDGES];
int dis[NODES];
int main(){
for(int i = 0; i < NODES; i++){
    dis[i] = MAX;
}
for(int i = 0; i < EDGES; i++){
    v[i].clear();
}
int nodes, edges, start_node, end_node, dist;
cin >> nodes >> edges;
for(int i = 0; i < edges; i++){
    cin >> start_node >> end_node >> dist;
    v[i].push_back(start_node);
    v[i].push_back(end_node);
    v[i].push_back(dist);
}
dis[1] = 0; //Assuming the source node is 1.
for(int i = 0; i < nodes-1; i++){
    int j = 0;
    while(v[j].size() != 0){
        if(dis[v[j][0]]+v[j][2]<dis[v[j][1]]){
            dis[v[j][1]] = dis[v[j][0]]+v[j][2];
        }
        j++;
    }
}
for(int i = 1; i <= nodes; i++){
    cout<< dis[i] << "\n";
}
return 0;
}

/*
Test Case:
    Input:
        5 5
        1 2 5
        1 3 2
        3 4 1
        1 4 6
        3 5 5

    Output: 
        5 2 3 7
*/